dkc/core/tests/test_authorized_uploads.py

```diff
@@ -0,0 +1,179 @@
+from datetime import timedelta
+
+from django.utils import timezone
+import pytest
+
+from dkc.core.models import AuthorizedUpload
+
+
+@pytest.fixture
+def authorized_upload(admin_api_client, folder):
+    return admin_api_client.post('/api/v2/authorized_uploads', data={'folder': folder.id})
+
+
+@pytest.mark.django_db
+def test_create_authorized_upload(admin_api_client, folder):
+    resp = admin_api_client.post('/api/v2/authorized_uploads', data={'folder': folder.id})
+    assert resp.status_code == 201
+    assert resp.data['folder'] == folder.id
+    assert resp.data['creator'] == admin_api_client.handler._force_user.id
+    assert 'signature' in resp.data
+
+
+@pytest.mark.django_db
+def test_authorized_upload_requires_write_access(api_client, public_folder, user):
+    api_client.force_authenticate(user)
+    resp = api_client.post('/api/v2/authorized_uploads', data={'folder': public_folder.id})
+    assert resp.status_code == 403
+
+
+@pytest.mark.django_db
+def test_authorized_upload_signature_check(authorized_upload, api_client):
+    resp = api_client.post(
+        '/api/v2/files',
+        data={
+            'name': 'foo.txt',
+            'size': 123,
+            'folder': authorized_upload.data['folder'],
+            'authorization': f'1{authorized_upload.data["signature"]}',
+        },
+    )
+    assert resp.status_code == 403
+    assert resp.json() == {'detail': 'Invalid authorization signature.'}
+
+
+@pytest.mark.django_db
+def test_authorized_upload_folder_check(authorized_upload, api_client, folder_factory):
+    folder2 = folder_factory()
+    resp = api_client.post(
+        '/api/v2/files',
+        data={
+            'name': 'foo.txt',
+            'size': 123,
+            'folder': folder2.id,
+            'authorization': authorized_upload.data['signature'],
+        },
+    )
+    assert resp.status_code == 403
+    assert resp.json() == {'detail': 'This upload was authorized to a different folder.'}
+
+
+@pytest.mark.django_db
+def test_authorized_upload_expired(api_client, folder, user_factory):
+    admin = user_factory(is_superuser=True)
+    authorized_upload = AuthorizedUpload.objects.create(
+        creator=admin,
+        folder=folder,
+        expires=timezone.now() - timedelta(hours=1),
```

Comment:
This may become moot if we change the implementation, but I think it's better to initialize `created` in the past, rather than try to mess with `expires`. That should result in a more realistic instance, making this test more robust to future implementation changes.

Of course (especially if we're using a Factory), we need to make sure that this isn't swallowed by `CreationDateTimeField` (which does always cooperate with being manually set).

Reply:
I changed this as a result of making `expires` computed. Unfortunately we can't override the `auto_now_add` behavior of the `created` field at creation time, so I am just updating it afterward for this test case.